### PR TITLE
Apply date passed from Event Manager to event start date. [ECP-954]

### PR DESCRIPTION
**Pull request by EugeneKyale**
This PR ensures the date selected when creating a new event from the Event Manager is applied to the block editor.

Ticket : https://theeventscalendar.atlassian.net/browse/ECP-954
Artifact : https://www.loom.com/share/59f795e3736c4f7ba12319ee73bdd685
Updated Artifact : https://www.loom.com/share/b2c7e4f13be74c42b3ab56a71c3f179b

### CODE REVIEW

File: src/Tribe/Editor/Objects/Event.php [hunk 1]
```diff
@@ -59,6 +59,17 @@ public function data( $key = null, $default = null ) {
...
+			/**
+			 * Grabs the tribe-start-date query param from the url 
+			 * if it exists then adds it to the global window object.
+			 */
+			if ( isset( $_GET['tribe-start-date'] ) ) {
+				$this->data = [
+					'is_new_post'      => true,
+					'tribe_start_date' => $_GET['tribe-start-date'],
+				];
+			}
```
Comment (paulmskim) [inline src/Tribe/Editor/Objects/Event.php:71]: Getting data from a user-input data is inherently unsafe, and using it directly in our application is also unsafe. A few things:

1. There's a function called `tribe_get_request_var()` (see common for the function definition) that allows us to get the variable from `$_GET` safely by sanitizing the value, use that.
2. We should be checking if the `tribe-start-date` we get is of proper format. If not, we shouldn't output it onto the data.
3. No need to redefine `$this->data` here, just define a new key and value.

File: src/modules/data/blocks/datetime/reducer.js [hunk 2]
```diff
@@ -27,9 +27,10 @@ export const defaultEndMoment = moment().startOf( 'day' ).seconds( defaultEndTim
 
 const defaultStartDateTime = momentUtil.toDateTime( defaultStartMoment );
 const defaultEndDateTime = momentUtil.toDateTime( defaultEndMoment );
+const queryStartDate = globals.postObjects().tribe_events.tribe_start_date;
 
 export const DEFAULT_STATE = {
-	start: defaultStartDateTime,
+	start: queryStartDate ? queryStartDate : defaultStartDateTime,
```
Comment (paulmskim) [inline src/modules/data/blocks/datetime/reducer.js:33]: Doing this replaces the default start time. In your screencast, the date is correct, but it is no longer the 8:00 AM that it used to be.

**Review by paulmskim — CHANGES_REQUESTED** [review]

File: src/modules/data/blocks/datetime/reducer.js (hunk 2 above)
Comment (EugeneKyale) [reply src/modules/data/blocks/datetime/reducer.js:33]: I've now included the correct start time.

File: src/Tribe/Editor/Objects/Event.php (hunk 1 above)
Comment (EugeneKyale) [reply src/Tribe/Editor/Objects/Event.php]: I'm now using the `tribe_get_request_var()` function to get the variable and validating it before adding it to the data.

**Review by bordoni — DISMISSED** [review]

**Review by paulmskim — CHANGES_REQUESTED** [review]

File: src/modules/data/blocks/datetime/reducer.js (hunk 2 above)
Comment (EugeneKyale) [reply src/modules/data/blocks/datetime/reducer.js]: @paulmskim I've made an update here, please let me know if that's correct.

File: src/Tribe/Editor/Objects/Event.php [hunk 3]
```diff
@@ -59,6 +59,19 @@ public function data( $key = null, $default = null ) {
...
+			 * Grabs the tribe-start-date query param from the url 
+			 * if it exists then adds it to the global window object.
+			 */
+			if ( $startDate && strtotime( $startDate ) ) {
+				$this->data = [
+					'is_new_post'      => true,
+					'tribe_start_date' => $startDate . ' 08:00:00',
+				];
+			}
+
```
[the remark this answers is not in the document]
Comment (EugeneKyale) [reply src/Tribe/Editor/Objects/Event.php]: @bordoni I've made an update here, please let me know if that's correct.

File: src/modules/data/blocks/datetime/reducer.js [hunk 4]
```diff
@@ -27,9 +27,11 @@ export const defaultEndMoment = moment().startOf( 'day' ).seconds( defaultEndTim
 
 const defaultStartDateTime = momentUtil.toDateTime( defaultStartMoment );
 const defaultEndDateTime = momentUtil.toDateTime( defaultEndMoment );
+const queryStartDate = globals.postObjects().tribe_events.tribe_start_date;
+const queryStartDateTime = queryStartDate + ' ' + defaultStartTime;
 
 export const DEFAULT_STATE = {
-	start: defaultStartDateTime,
+	start: queryStartDate ? queryStartDateTime : defaultStartDateTime,
```
Comment (paulmskim) [inline src/modules/data/blocks/datetime/reducer.js:34]: set line 25 above:

```
const queryStartDate = globals.postObjects().tribe_events.tribe_start_date;
export const defaultStartMoment = queryStartDate ? moment( queryStartDate ).seconds( defaultStartTimeSeconds ) : moment().startOf( 'day' ).seconds( defaultStartTimeSeconds );
```

Remove the rest of the additions here, the `defaultStartMoment` creates the `defaultStartDateTime`, which sets the `DEFAULT_STATE`.

File: src/Tribe/Editor/Objects/Event.php [hunk 5]
```diff
@@ -59,6 +59,19 @@ public function data( $key = null, $default = null ) {
...
+			/**
+			 * Grabs the tribe-start-date query param from the url 
+			 * if it exists then adds it to the global window object.
+			 */
+			if ( $startDate && strtotime( $startDate ) ) {
+				$this->data = [
+					'is_new_post'      => true,
+					'tribe_start_date' => $startDate . ' 08:00:00',
+				];
```
[the remark this answers is not in the document]
Comment (EugeneKyale) [reply src/Tribe/Editor/Objects/Event.php]: Ooh, I see. I've updated made the update.

File: src/modules/data/blocks/datetime/reducer.js [hunk 6]
```diff
@@ -21,8 +21,9 @@ const defaultStartTime = globals.defaultTimes().start ? globals.defaultTimes().s
 const defaultEndTime = globals.defaultTimes().end ? globals.defaultTimes().end : '17:00:00';
 const defaultStartTimeSeconds = time.toSeconds( defaultStartTime, time.TIME_FORMAT_HH_MM_SS );
 const defaultEndTimeSeconds = time.toSeconds( defaultEndTime, time.TIME_FORMAT_HH_MM_SS );
+const queryStartDate = globals.postObjects().tribe_events.tribe_start_date;
 
-export const defaultStartMoment = moment().startOf( 'day' ).seconds( defaultStartTimeSeconds );
+export const defaultStartMoment = queryStartDate ? moment( queryStartDate ).seconds( defaultStartTimeSeconds ) : moment().startOf( 'day' ).seconds( defaultStartTimeSeconds ); // eslint-disable-line max-len
```
Comment (paulmskim) [inline src/modules/data/blocks/datetime/reducer.js:26]: Use `// eslint-disable-line` as a last resort if you cannot break code into multiple lines (for example, a long string of text that cannot be broken into 2 lines). This, you can break into multiple lines:

```suggestion
export const defaultStartMoment = queryStartDate
	? moment( queryStartDate ).seconds( defaultStartTimeSeconds )
	: moment().startOf( 'day' ).seconds( defaultStartTimeSeconds );
```

File: src/Tribe/Editor/Objects/Event.php [hunk 7]
```diff
@@ -59,6 +60,16 @@ public function data( $key = null, $default = null ) {
...
 
+			$startDate = tribe_get_request_var( 'tribe-start-date' );
+
+			/**
+			 * Grabs the tribe-start-date query param from the url 
+			 * if it exists then adds it to the global window object.
+			 */
+			if ( $startDate && strtotime( $startDate ) ) {
+				$this->data['tribe_start_date'] = Dates::build_date_object( $startDate )->format( Dates::DBDATEFORMAT );
+			}
```
Comment (paulmskim) [inline src/Tribe/Editor/Objects/Event.php:71]: This is close, but best to follow the implementation that we used in the classic editor (see https://github.com/the-events-calendar/the-events-calendar/blob/master/src/Tribe/Admin/Event_Meta_Box.php#L96).

**Review by paulmskim — CHANGES_REQUESTED** [review]
A couple more things and this is good to go

File: src/Tribe/Editor/Objects/Event.php (hunk 7 above)
Comment (EugeneKyale) [reply src/Tribe/Editor/Objects/Event.php]: Thanks @paulmskim I've modified this to use a similar implementation to the classic editor.

**Review by paulmskim — APPROVED** [review]

**Pull request merged**